cluster-autoscaler/expander/priority/priority.go

```diff
@@ -0,0 +1,189 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package priority
+
+import (
+	"fmt"
+	"regexp"
+	"sync"
+
+	"gopkg.in/yaml.v2"
+
+	"k8s.io/autoscaler/cluster-autoscaler/expander"
+	"k8s.io/autoscaler/cluster-autoscaler/expander/random"
+	"k8s.io/autoscaler/cluster-autoscaler/utils/errors"
+
+	apiv1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/klog"
+	schedulernodeinfo "k8s.io/kubernetes/pkg/scheduler/nodeinfo"
+)
+
+type priority struct {
+	fallbackStrategy expander.Strategy
+	changesChan      <-chan watch.Event
+	priorities       map[int][]*regexp.Regexp
+	padlock          sync.RWMutex
```

Comment:
Expander is called once per loop, at most 1 call / 10s (more likely 1 per 15-20s), after a ton of complex calculations done all over CA code. Do we really need async config update? Feels like we could just get configmap from watch cache every time and synchronously check if the config changed - the performance impact should be negligible and the complexity of this expander will greatly reduce.

Reply:
Hmm, I agree, if there's no plan to make the updates/loops more frequent in the future, probably we can drop the watch and go with fetching the map every time. I'm just not sure about caching - can you point me to some docs? How is stuff cached, for how long and what is your default caching toolset for CAS?

Reply:
By 'watch cache' I mean using listers or informers. Those are a higher level abstraction for Kubernetes client provided by client-go. They provide a local cache that is backed by a watch on apiserver and updated on any change. So you get fresh data, except you don't have to care about creating goroutines doing the watch and error handling - it's all provided.
The general recommendation is to always use lister or informer over raw client, unless there is a very good reason not to.